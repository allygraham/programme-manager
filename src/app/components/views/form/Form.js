import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';
import { constants } from '../../../constants';

const LabelText = styled('span')`
  margin-left: 0.5rem;
`;

const Button = styled('button')`
  margin-top: 1rem;
`;

const Notification = styled('article')`
  margin: 1rem 0;
`;

export class Form extends Component {

  constructor(props) {
    super(props);

    this.state = {
      id: props.formData.id,
      name: props.formData.name,
      shortDescription: props.formData.shortDescription,
      active: props.formData.active,
      error: '',
    };

    this.onSubmitForm = this.onSubmitForm.bind(this);
  }

  componentDidUpdate(prevProps) {
    if (this.props.formData.id !== prevProps.formData.id) {
      this.setState({
        id: this.props.formData.id,
        name: this.props.formData.name,
        shortDescription: this.props.formData.shortDescription,
        active: this.props.formData.active,
        error: '',
      });
    }
  }

  validateForm() {
    const { id, name, shortDescription } = this.state;

    if (!id || id === 0) {
      this.setState({
        error: 'You must specify a programme ID and it can\'t be 0!'
      });
      return true;
    }

    if (!name) {
      this.setState({
        error: 'You must specify a programme name!'
      });
      return true;
    }

    if (!shortDescription) {
      this.setState({
        error: 'You must specify a programme description!'
      });
      return true;
    }
  }

  onResetForm() {
    this.setState({
      id: 0,
      name: '',
      shortDescription: '',
      active: false,
      error: '',
    });
  }

  onSubmitForm() {
    const { onSubmit: submitForm } = this.props;
    const { id, name, shortDescription, active } = this.state;
    console.log('submitting the form');

    if (this.validateForm()) {
      return;
    }

    const programme = {
      id: id,
      name: name,
      shortDescription: shortDescription,
      active: active,
    };

    submitForm(programme);
    this.onResetForm();
  }

  render() {
    const { id, name, shortDescription, active, error } = this.state;
    const { isEditing } = this.props;

    return (
      <form>
        <div className="field">
          <label className="label">{constants.tableTitles.id}</label>
          <div className="control">
            <input
              className="input"
              type="number"
              placeholder="Enter an ID..."
              value={id}
              onChange={(event) => this.setState({ id: Number(event.target.value) })}
              disabled={isEditing}
            />
          </div>
        </div>

        <div className="field">
          <label className="label">{constants.tableTitles.name}</label>
          <div className="control">
            <input
              className="input"
              type="text"
              placeholder="Enter a name..."
              value={name}
              onChange={(event) => this.setState({ name: event.target.value })}
            />
          </div>
        </div>

        <div className="field">
          <label className="label">{constants.tableTitles.description}</label>
          <div className="control">
            <textarea
              className="textarea"
              placeholder="Enter a description..."
              value={shortDescription}
              onChange={(event) => this.setState({ shortDescription: event.target.value })}
            />
          </div>
        </div>

        <div className="field">
          <div className="control">
            <label className="checkbox">
              <input
                type="checkbox"
                checked={active}
                onChange={() => this.setState({ active: !this.state.active })}
              />
              <LabelText>{constants.form.checkboxLabel}</LabelText>
            </label>
          </div>
        </div>

        {error.length > 0 &&
          <Notification className="message is-danger">
            <div className="message-body">
              {error}
            </div>
          </Notification>
        }

        <div className="field">
          <div className="control">
            <Button
              className="button is-link"
              type="submit"
              onClick={this.onSubmitForm}
            >
              {constants.buttons.save}
            </Button>
          </div>
        </div>

      </form>
    );
  }
}

Form.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  formData: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    shortDescription: PropTypes.string,
    active: PropTypes.bool,
  }).isRequired,
  isEditing: PropTypes.bool.isRequired,
};
