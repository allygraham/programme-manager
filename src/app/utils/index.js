export const sortByName = (name1, name2) => {
  if (name1 < name2) {
    return -1;
  }
  if (name1 > name2) {
    return 1;
  }
  return 0;
};

export const sortById = (id1, id2) => {
  if (id1 < id2) {
    return -1;
  }
  if (id1 > id2) {
    return 1;
  }
  return 0;
};

export const blankForm = {
  id: 0,
  name: '',
  shortDescription: '',
  active: false,
};
