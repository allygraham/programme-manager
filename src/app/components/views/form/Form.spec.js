import React from 'react';
import { shallow } from 'enzyme';

import { blankForm } from '../../../utils';
import { Form } from './Form';

describe('Form', () => {

  test('render the <Form> component', () => {
    const formData = { id: parseInt(blankForm.id), name: blankForm.name, shortDescription: blankForm.shortDescription, active: blankForm.active };
    const onSubmit = jest.fn();
    const isEditing = false;

    const wrapper = shallow(<Form formData={formData} onSubmit={onSubmit} isEditing={isEditing}/>);
    expect(wrapper).toMatchSnapshot();
  });

});