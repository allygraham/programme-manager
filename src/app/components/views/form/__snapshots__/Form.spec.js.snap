// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form render the <Form> component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    formData={
      Object {
        "active": false,
        "id": 0,
        "name": "",
        "shortDescription": "",
      }
    }
    isEditing={false}
    onSubmit={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="field"
        >
          <label
            className="label"
          >
            ID
          </label>
          <div
            className="control"
          >
            <input
              className="input"
              disabled={false}
              onChange={[Function]}
              placeholder="Enter an ID..."
              type="number"
              value={0}
            />
          </div>
        </div>,
        <div
          className="field"
        >
          <label
            className="label"
          >
            Name
          </label>
          <div
            className="control"
          >
            <input
              className="input"
              onChange={[Function]}
              placeholder="Enter a name..."
              type="text"
              value=""
            />
          </div>
        </div>,
        <div
          className="field"
        >
          <label
            className="label"
          >
            Description
          </label>
          <div
            className="control"
          >
            <textarea
              className="textarea"
              onChange={[Function]}
              placeholder="Enter a description..."
              value=""
            />
          </div>
        </div>,
        <div
          className="field"
        >
          <div
            className="control"
          >
            <label
              className="checkbox"
            >
              <input
                checked={false}
                onChange={[Function]}
                type="checkbox"
              />
              <Styled(span)>
                Is this programme currently active?
              </Styled(span)>
            </label>
          </div>
        </div>,
        false,
        <div
          className="field"
        >
          <div
            className="control"
          >
            <Styled(button)
              className="button is-link"
              onClick={[Function]}
              type="submit"
            >
              Save
            </Styled(button)>
          </div>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label
              className="label"
            >
              ID
            </label>,
            <div
              className="control"
            >
              <input
                className="input"
                disabled={false}
                onChange={[Function]}
                placeholder="Enter an ID..."
                type="number"
                value={0}
              />
            </div>,
          ],
          "className": "field",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "ID",
              "className": "label",
            },
            "ref": null,
            "rendered": "ID",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                className="input"
                disabled={false}
                onChange={[Function]}
                placeholder="Enter an ID..."
                type="number"
                value={0}
              />,
              "className": "control",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "input",
                "disabled": false,
                "onChange": [Function],
                "placeholder": "Enter an ID...",
                "type": "number",
                "value": 0,
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label
              className="label"
            >
              Name
            </label>,
            <div
              className="control"
            >
              <input
                className="input"
                onChange={[Function]}
                placeholder="Enter a name..."
                type="text"
                value=""
              />
            </div>,
          ],
          "className": "field",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Name",
              "className": "label",
            },
            "ref": null,
            "rendered": "Name",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                className="input"
                onChange={[Function]}
                placeholder="Enter a name..."
                type="text"
                value=""
              />,
              "className": "control",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "input",
                "onChange": [Function],
                "placeholder": "Enter a name...",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label
              className="label"
            >
              Description
            </label>,
            <div
              className="control"
            >
              <textarea
                className="textarea"
                onChange={[Function]}
                placeholder="Enter a description..."
                value=""
              />
            </div>,
          ],
          "className": "field",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Description",
              "className": "label",
            },
            "ref": null,
            "rendered": "Description",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <textarea
                className="textarea"
                onChange={[Function]}
                placeholder="Enter a description..."
                value=""
              />,
              "className": "control",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "textarea",
                "onChange": [Function],
                "placeholder": "Enter a description...",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "textarea",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="control"
          >
            <label
              className="checkbox"
            >
              <input
                checked={false}
                onChange={[Function]}
                type="checkbox"
              />
              <Styled(span)>
                Is this programme currently active?
              </Styled(span)>
            </label>
          </div>,
          "className": "field",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <label
              className="checkbox"
            >
              <input
                checked={false}
                onChange={[Function]}
                type="checkbox"
              />
              <Styled(span)>
                Is this programme currently active?
              </Styled(span)>
            </label>,
            "className": "control",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  checked={false}
                  onChange={[Function]}
                  type="checkbox"
                />,
                <Styled(span)>
                  Is this programme currently active?
                </Styled(span)>,
              ],
              "className": "checkbox",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "checked": false,
                  "onChange": [Function],
                  "type": "checkbox",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Is this programme currently active?",
                },
                "ref": null,
                "rendered": "Is this programme currently active?",
                "type": [Function],
              },
            ],
            "type": "label",
          },
          "type": "div",
        },
        "type": "div",
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="control"
          >
            <Styled(button)
              className="button is-link"
              onClick={[Function]}
              type="submit"
            >
              Save
            </Styled(button)>
          </div>,
          "className": "field",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Styled(button)
              className="button is-link"
              onClick={[Function]}
              type="submit"
            >
              Save
            </Styled(button)>,
            "className": "control",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Save",
              "className": "button is-link",
              "onClick": [Function],
              "type": "submit",
            },
            "ref": null,
            "rendered": "Save",
            "type": [Function],
          },
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="field"
          >
            <label
              className="label"
            >
              ID
            </label>
            <div
              className="control"
            >
              <input
                className="input"
                disabled={false}
                onChange={[Function]}
                placeholder="Enter an ID..."
                type="number"
                value={0}
              />
            </div>
          </div>,
          <div
            className="field"
          >
            <label
              className="label"
            >
              Name
            </label>
            <div
              className="control"
            >
              <input
                className="input"
                onChange={[Function]}
                placeholder="Enter a name..."
                type="text"
                value=""
              />
            </div>
          </div>,
          <div
            className="field"
          >
            <label
              className="label"
            >
              Description
            </label>
            <div
              className="control"
            >
              <textarea
                className="textarea"
                onChange={[Function]}
                placeholder="Enter a description..."
                value=""
              />
            </div>
          </div>,
          <div
            className="field"
          >
            <div
              className="control"
            >
              <label
                className="checkbox"
              >
                <input
                  checked={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <Styled(span)>
                  Is this programme currently active?
                </Styled(span)>
              </label>
            </div>
          </div>,
          false,
          <div
            className="field"
          >
            <div
              className="control"
            >
              <Styled(button)
                className="button is-link"
                onClick={[Function]}
                type="submit"
              >
                Save
              </Styled(button)>
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label
                className="label"
              >
                ID
              </label>,
              <div
                className="control"
              >
                <input
                  className="input"
                  disabled={false}
                  onChange={[Function]}
                  placeholder="Enter an ID..."
                  type="number"
                  value={0}
                />
              </div>,
            ],
            "className": "field",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "ID",
                "className": "label",
              },
              "ref": null,
              "rendered": "ID",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <input
                  className="input"
                  disabled={false}
                  onChange={[Function]}
                  placeholder="Enter an ID..."
                  type="number"
                  value={0}
                />,
                "className": "control",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "input",
                  "disabled": false,
                  "onChange": [Function],
                  "placeholder": "Enter an ID...",
                  "type": "number",
                  "value": 0,
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label
                className="label"
              >
                Name
              </label>,
              <div
                className="control"
              >
                <input
                  className="input"
                  onChange={[Function]}
                  placeholder="Enter a name..."
                  type="text"
                  value=""
                />
              </div>,
            ],
            "className": "field",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Name",
                "className": "label",
              },
              "ref": null,
              "rendered": "Name",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <input
                  className="input"
                  onChange={[Function]}
                  placeholder="Enter a name..."
                  type="text"
                  value=""
                />,
                "className": "control",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "input",
                  "onChange": [Function],
                  "placeholder": "Enter a name...",
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label
                className="label"
              >
                Description
              </label>,
              <div
                className="control"
              >
                <textarea
                  className="textarea"
                  onChange={[Function]}
                  placeholder="Enter a description..."
                  value=""
                />
              </div>,
            ],
            "className": "field",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Description",
                "className": "label",
              },
              "ref": null,
              "rendered": "Description",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <textarea
                  className="textarea"
                  onChange={[Function]}
                  placeholder="Enter a description..."
                  value=""
                />,
                "className": "control",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "textarea",
                  "onChange": [Function],
                  "placeholder": "Enter a description...",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "textarea",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="control"
            >
              <label
                className="checkbox"
              >
                <input
                  checked={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <Styled(span)>
                  Is this programme currently active?
                </Styled(span)>
              </label>
            </div>,
            "className": "field",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <label
                className="checkbox"
              >
                <input
                  checked={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <Styled(span)>
                  Is this programme currently active?
                </Styled(span)>
              </label>,
              "className": "control",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    checked={false}
                    onChange={[Function]}
                    type="checkbox"
                  />,
                  <Styled(span)>
                    Is this programme currently active?
                  </Styled(span)>,
                ],
                "className": "checkbox",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "checked": false,
                    "onChange": [Function],
                    "type": "checkbox",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Is this programme currently active?",
                  },
                  "ref": null,
                  "rendered": "Is this programme currently active?",
                  "type": [Function],
                },
              ],
              "type": "label",
            },
            "type": "div",
          },
          "type": "div",
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="control"
            >
              <Styled(button)
                className="button is-link"
                onClick={[Function]}
                type="submit"
              >
                Save
              </Styled(button)>
            </div>,
            "className": "field",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Styled(button)
                className="button is-link"
                onClick={[Function]}
                type="submit"
              >
                Save
              </Styled(button)>,
              "className": "control",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Save",
                "className": "button is-link",
                "onClick": [Function],
                "type": "submit",
              },
              "ref": null,
              "rendered": "Save",
              "type": [Function],
            },
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
